platforms/nuttx/src/px4/common/usr_hrt.cpp

```diff
@@ -80,9 +80,7 @@ hrt_absolute_time(void)
 void
 hrt_store_absolute_time(volatile hrt_abstime *t)
 {
-	irqstate_t flags = px4_enter_critical_section();
```

Comment:
The store is intended to be atomic.

Reply:
Thanks for the review and comment. Yes, the handling of the timer needs to be atomic, this is handled by the kernel side hrt driver. Let me try to explain:

In CONFIG_BUILD_PROTECTED i.e. px4 built into separate kernel-/userspace, this module (usr_hrt) is the userspace hrt "driver". The userspace driver itself does not contain any functionality besides `boardctl()` calls into the actual hrt driver, that exists in kernel space. The kernel space driver handles the actual driver hardware access, as well as atomicy regarding this access.

The userspace version of `hrt_store_absolute_time()` actually just calls the userspace version of `hrt_absolute_time()` which uses a call gate to enter kernel mode, which handles the timer access.

```
hrt_abstime
hrt_absolute_time(void)
{
	hrt_abstime abstime = 0;
	boardctl(HRT_ABSOLUTE_TIME, (uintptr_t)&abstime);
	return abstime;
}
```

So the atomicy is not necessary in user space. Also, for me using `px4_enter_critical_section()` from unprivileged (user) mode results in a full system crash due to an illegal instruction exception. My host target is RISC-V which does not allow disabling global interrupts from user code.

So why does this work for ARM then ? Well, it sort of does and doesn't.

According to the ARM documentation:
_Use CPS only from privileged software. It has no effect if used in unprivileged software._
I used the term “sort of”, because interrupts are not masked in places where this is expected, but no exception is triggered either. Instead the instruction is eaten away silently. More seriously, depending on a global interrupt disable as a synchronization barrier in userspace can cause nasty race / sync issues which will be impossible to debug without knowing the quoted line from the ARM documentation.

Hopefully I explained the issue clear enough. If something is not clear I will be more than happy to explain it!

Reply:
Thanks for the explanation, and I agree, I'd just think about the bigger picture where `hrt_store_absolute_time()` is only used for convenience in a couple places. 

So what if we just get rid of hrt_store_absolute_time for user space?

Reply:
Ok, I can look into that. I don't have a very thorough understanding of the modules that were duplicated / implemented for user space, I just made a specific fix for the full system crash. 

Thankfully the one who implemented usr_hrt sits next to me so I can ask him directly the motivation for it and maybe get to know the dependencies better.

Reply:
I think @pussuw 's analysis is correct, there is no point in trying to enter critical section in any of the usr_hrt function, since those are all just ioctls to kernel side, and the critical section (if needed) is handled there. It was originally a mistake from me, I didn't think this through well enough. So those irq-disabling calls are actually all no-ops in arm when executed from user-space, and trigger crash on risc-v. The fix is correct, they should be removed in usr_hrt. Of course it doesn't harm to get rid of hrt_store_absolute_time for user space entirely. But, removing the hrt_store_absolute_time from usr_hrt causes that you just can't put any module calling to user space any more - a limitation which might be unnecessary.

Reply:
We could purge hrt_store_absolute_time entirely and replace the 1 or 2 uses (our nuttx cpuload code) uses with critical sections directly. The handful of other uses I see (paw3902, etc) seem to be using it just for convenience/laziness.

Reply:
Simplifying the hrt interface by removing that sounds very good. Could you @dagar clarify whether you'd like the author of this PR to do that or are you planning to do that yourself?

Or would it make sense to first just remove the critical section from the usr_hrt, which is proposed in this PR. It is obvious that the critical sections in the usr_hrt is just a bug (even though they are no-ops in arm builds). So this PR is correct IMHO.

Reply:
I removed the contested method entirely -> resolving this